### PR TITLE
First pass at matrix stuff!

### PR DESCRIPTION
**Pull request by dcdillon**
Big change.  Instead of only allowing `Vector` results, results now have `dims` which are two dimensional.  This means that every function/functor pair we create needs to be able to calculate the `dims` of its result.

I still haven't quite thought of a way to handle things like matrix multiplication, however, I suspect that it will be a parameter passed to functions to indicate whether things are row or column major.

So right now, what we have here, is something that can operate memberwise on any `Matrix` or `Vector`.  Not bad.

### CODE REVIEW

**Comment by dcdillon** [comment]
Fix for `wrap` forthcoming and then this should be a go.

**Comment by dcdillon** [comment]
This will fail until RcppCore/Rcpp#585 is merged to fix RcppCore/Rcpp#584 and the new release is out.

**Comment by dcdillon** [comment]
Closing this until the new Rcpp release.

File: .travis.yml [hunk 1]
```diff
@@ -12,7 +12,8 @@ before_install:
   - ./run.sh bootstrap
 
 install:
-  - ./run.sh install_aptget r-cran-rcpp r-cran-runit r-cran-knitr r-cran-rmarkdown
+  - ./run.sh install_r Rcpp
```
Comment (eddelbuettel) [inline .travis.yml:15]: I _think_ can also install this via apt-get as the Rutter PPAs (which `run.sh` looks at) has current versions.

At least I never install from source.

File: R/plugin.R [hunk 2]
```diff
@@ -1 +1,18 @@
...
+    shlib_ld <- system("R CMD config SHLIB_LD", intern=TRUE)
+    shlib_cxxld <- system("R CMD config SHLIB_CXXLD", intern=TRUE)
+
+    shlib_ld_results <- system(paste0(shlib_ld, " --version"), intern=TRUE)
+    shlib_cxxld <- system(paste0(shlib_cxxld, " --version"), intern=TRUE)
+
+    shlib_ld_good <- any(grepl(linker_pattern, shlib_ld_results))
+    shlib_cxxld_good <- any(grepl(linker_pattern, shlib_cxxld_results))
+
+    libs <- (shlib_ld_good && shlib_cxxld_good) "-Wl,-S" else ""
```
Comment (eddelbuettel) [inline R/plugin.R]: That lack an `if``.  Can't parse as is.

**Comment by dcdillon** [comment]
Should be fixed now

**Pull request merged**